hooks/useLocalStorage.tsx

```diff
@@ -1,17 +1,57 @@
-import { useEffect, useState } from 'react';
+import { useEffect, useCallback, useState, Dispatch, SetStateAction } from 'react';
 
-export default function useLocalStorage(key: string, initial: any) {
-    const [value, setValue] = useState(() => {
-        if (typeof window !== 'undefined') {
-            const data = window.localStorage.getItem(key);
-            if (data !== null) return JSON.parse(data);
+export default function useLocalStorage<T>(key: string, initialValue: T, restoreOnDelete: boolean = false): [T, Dispatch<SetStateAction<T | undefined>>, () => void] {
+    const deserializer = JSON.parse;
+    const serializer = JSON.stringify;
+
+    const [state, setState] = useState<T>(initialValue);
+
+    useEffect(() => {
+        if (process.browser) {
+            const localStorageValue = localStorage.getItem(key);
+            if (localStorageValue !== null) {
+                setState(deserializer(localStorageValue));
+            } else {
+                localStorage.setItem(key, serializer(initialValue));
+                setState(initialValue);
+            }
+        }
+    }, [key, process.browser]);
```

Comment:
It doesn't look as it should thanks to the pre-rendering.
A small solution was to do this